aiosmtpd/docs/smtp.rst

```diff
@@ -104,31 +107,82 @@ override to provide additional responses.
-or in exceptional cases.  More hooks may be added in the future, but for now
-the following handler hooks are defined:
+Handlers can implement a hooks that get called during the ``SMTP`` dialog, or
+in exceptional cases.  These *handler hooks* are all called asynchronously and
+they *must* return a status string, since the default statuses are *not*
+returned when a hook is defined.  Individual handlers may have additional
+responsibilities to replace default behavior, as described below.
+
+All handler hooks take at least three arguments, the ``SMTP`` server instance,
+:ref:`a session instance, and an envelope instance <sessions_and_envelopes>`.
+Some methods take additional arguments.
+
+The following hooks are currently defined:
+
+``handle_HELO(server, session, envelope, hostname)``
+    Called during ``HELO``, the ``hostname`` argument is the host name given
+    by the client in the ``HELO`` command.  If implemented, this hook must
+    also set the ``session.host_name`` attribute.
+
+``handle_EHLO(server, session, envelope, hostname)``
+    Called during ``EHLO``, the ``hostname`` argument is the host name given
+    by the client in the ``EHLO`` command.  If implemented, this hook must
+    also set the ``session.host_name`` attribute.  This hook may push
+    additional ``250-<command>`` responses to the client by yielding from
+    ``server.push(status)``.
+
+``handle_NOOP(server, session, envelope)``
+    Called during ``NOOP``.
+
+``handle_QUIT(server, session, envelope)``
+    Called during ``QUIT``.
+
+``handle_VRFY(server, session, envelope, address)``
+    Called during ``VRFY``, the ``address`` argument is the parsed email
+    address given by the client in the ``VRFY`` command.
+
+``handle_MAIL(server, session, envelope, address, mail_options)``
+    Called during ``MAIL FROM``, the ``address`` argument is the parsed email
+    address given by the client in the ``MAIL FROM`` command, and
+    ``mail_options`` are any additional ESMTP mail options providing by the
+    client.  If implemented, this hook must also set the
+    ``envelope.mail_from`` attribute and it may extend
+    ``envelope.mail_options`` (which is always a Python list).
+
+``handle_RCPT(server, session, envelope, address, rcpt_options)``
+    Called during ``RCPT TO``, the ``address`` argument is the parsed email
+    address given by the client in the ``RCPT TO`` command, and
+    ``rcpt_options`` are any additional ESMTP recipient options providing by
+    the client.  If implemented, this hook should append the address to
+    ``envelope.rcpt_tos`` and may extend ``envelope.rcpt_options`` (both of
+    which are always Python lists).
+
+``handle_RSET(server, session, envelope)``
+    Called during ``RSET``.
+
+``handle_DATA(session, envelope)``
+    Called during ``DATA`` after most processing of the data has occurred.
+    Hooks can inspect or change the converted data by looking at
+    ``envelope.content``.
```

Comment:
I think "after most processing of the data has occurred" sounds very vague. I suggest changing the first sentence to "Called during ``DATA`` when the entire message has been received."

Also, in what sense is `envelope.content` the *converted* data? Wouldn't it be more precise to call this the *incoming* data?